### PR TITLE
Add unit tests for all expressions and statements

### PR DESCRIPTION
**Pull request by ostlerc**
There is quite a bit of work here. But the general idea is I made a helper class called ProgramFactory that chains calls actling like parsed lines of cpsl. Then there are helper functions to either print out the mips of any one expression, or the list of built up statements.

### CODE REVIEW

**Comment by ostlerc** [comment]
Here is the coverage info I have that applies with this pull request
screen shot 2015-02-25 at 6 37 45 pm

File: AST_INC/AST/BasicBlock.cpp [hunk 1]
```diff
@@ -4,11 +4,20 @@
 #include <algorithm>
 #include <iterator>
 
-std::string cs6300::BasicBlock::getLabel()
+std::string cs6300::BasicBlock::getNextLabel(bool reset)
```
Comment (ksundberg) [inline AST_INC/AST/BasicBlock.cpp]: I don't like this.  I'd rather have a Label Factory with a get and a separate reset function.  With this change the getNextLabel function has two very different uses with a easily misunderstood parameter to distinguish between them.

File: AST_INC/AST/Expressions/AdditionExpression.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ bool cs6300::AdditionExpression::isConst() const
 
 std::string cs6300::AdditionExpression::name() const
 {
-  return "\"+\"";
+  return R"("+")";
```
Comment (ksundberg) [inline AST_INC/AST/Expressions/AdditionExpression.cpp]: Raw string literals for the win :)

File: AST_INC/AST/Expressions/Expression.cpp [hunk 3]
```diff
@@ -19,9 +19,14 @@ std::string cs6300::Expression::node() const
   return id() + " [label=" + name() + "]";
 }
 
-int cs6300::Expression::getNextLabel()
+int cs6300::Expression::getNextLabel(bool reset)
```
Comment (ksundberg) [inline AST_INC/AST/Expressions/Expression.cpp]: Same comment as for the basic block labels.  This is a good canidate for a refactor between both of them I think.

File: AST_INC/AST/BasicBlock.cpp (hunk 1 above)
Comment (ostlerc) [inline AST_INC/AST/BasicBlock.cpp]: yeah... this was kind of a hack i'll agree.

**Pull request merged**